File: tests/native/src/js/objects.rs
use neon::vm::{Call, JsResult};
use neon::mem::Handle;
use neon::js::{JsNumber, JsString, JsObject, Object};

pub fn return_js_object(call: Call) -> JsResult<JsObject> {
    Ok(JsObject::new(call.scope))
}

pub fn return_js_object_with_mixed_content(call: Call) -> JsResult<JsObject> {
    let scope = call.scope;
    let js_object: Handle<JsObject> = JsObject::new(scope);
    js_object.set("number", JsNumber::new(scope, 9000))?;
    js_object.set("string", JsString::new(scope, "hello node").unwrap())?;
    Ok(js_object)
}

pub fn return_js_object_with_integer(call: Call) -> JsResult<JsObject> {
    let scope = call.scope;
    let js_object: Handle<JsObject> = JsObject::new(scope);
    js_object.set("number", JsNumber::new(scope, 9000))?;
    Ok(js_object)
}

pub fn return_js_object_with_string(call: Call) -> JsResult<JsObject> {
    let scope = call.scope;
    let js_object: Handle<JsObject> = JsObject::new(scope);
    js_object.set("string", JsString::new(scope, "hello node").unwrap())?;
    Ok(js_object)
}

File: test/dynamic/lib/objects.js
var addon = require('../native');
var assert = require('chai').assert;

describe('JsObject', function() {
  it('return the v8::Global object', function () {
      assert(global === addon.return_js_global_object());
  });

  it('return a JsObject built in Rust', function () {
    assert.deepEqual({}, addon.return_js_object());
  });

  it('return a JsObject with a number key value pair', function () {
    assert.deepEqual({number: 9000}, addon.return_js_object_with_number());
  });

  it('return a JsObject with an string key value pair', function () {
    assert.deepEqual({string: "hello node"}, addon.return_js_object_with_string());
  });

  it('return a JsObject with mixed content key value pairs', function () {
    assert.deepEqual({number: 9000, string: 'hello node'}, addon.return_js_object_with_mixed_content());
  });
});
